File: go/day1/main.go
package main

import (	
	input "aoc2020/inpututils"

	"fmt"
	"errors"
)

func main(){
	fmt.Println("--- Part One ---")
	fmt.Println(Part1("input.txt"))
	
	fmt.Println("--- Part Two ---")
	fmt.Println(Part2("input.txt"))
}

type set map[int]struct{}

func (s set) add(n int){
	s[n] = struct{}{}
}

func (s set) has(n int) bool{
	_, ok := s[n]
	return ok
}

func Part1(filename string) int {
	nums := input.ReadNumbers(filename)
	multiplied, err := twoSum(nums, 2020)
	if err != nil {
		return -1
	}
	return multiplied
}

func Part2(filename string) int {
	nums := input.ReadNumbers(filename)
	for i := 0; i < len(nums) - 2; i += 1 {
		num := nums[i]
		multiplier, err := twoSum(nums[i+1:], 2020 - num)
		if err == nil {
			return multiplier * num
		}
	}
	return -1
}

func twoSum(ints []int, target int) (int, error) {
	seen := set{}
	for _, i := range(ints){
		if seen.has(target - i){
			return (target - i) * i, nil
		}
		seen.add(i)
	}
	return -1, errors.New("No solution")
}

func check(err error) {
	if err != nil {
		panic(err)	
	}
}

File: python/src/day1/main.py
from src.common.file_utils import get_path, read_integers


def part_one(filename: str) -> int:
    ints = read_integers(get_path(__file__, filename))
    return two_sum(ints, 2020)


def part_two(filename: str) -> int:
    ints = read_integers(get_path(__file__, filename))
    # use twosum solution from p1 to reduce time complexity from
    # O(n^3) to O(n^2)
    for i, n in enumerate(ints):
        multiplier = two_sum(ints[i + 1:], 2020 - n)
        if multiplier != -1:
            return n * multiplier
    return -1


def two_sum(ints: list, target: int) -> int:
    seen = set()
    for i in ints:
        if target - i in seen:
            return i * (target - i)
        seen.add(i)
    return -1


if __name__ == '__main__':
    print("---Part One---")
    part_one_res = part_one("input.txt")
    print(part_one_res)

    print("---Part Two---")
    part_two_res = part_two("input.txt")
    print(part_two_res)
